From a8b4dce7b53d73de372e150028c970ee0a2a2e97 Mon Sep 17 00:00:00 2001
From: "Dmitry V. Levin" <ldv@altlinux.org>
Date: Mon, 16 Mar 2020 21:02:18 +0000
Subject: [PATCH] Introduce pam_inline.h

Introduce a new internal header file for definitions of handly inline
functions and macros providing some convenient functionality to libpam
and its modules.

* libpam/include/pam_cc_compat.h (PAM_SAME_TYPE): New macro.
* libpam/include/pam_inline.h: New file.
* libpam/Makefile.am (noinst_HEADERS): Add include/pam_inline.h.

Backported-by: Sai Harsha <ssuryad@amazon.com>
---
 libpam/Makefile.am                    |  2 +-
 libpam/include/pam_cc_compat.h        | 13 ++++++++++
 libpam/include/pam_inline.h           | 37 +++++++++++++++++++++++++++
 modules/pam_namespace/pam_namespace.c |  1 +
 4 files changed, 52 insertions(+), 1 deletion(-)
 create mode 100644 libpam/include/pam_inline.h

diff --git a/libpam/Makefile.am b/libpam/Makefile.am
index b427329..4181331 100644
--- a/libpam/Makefile.am
+++ b/libpam/Makefile.am
@@ -19,7 +19,7 @@ include_HEADERS = include/security/_pam_compat.h \
 
 noinst_HEADERS = pam_prelude.h pam_private.h pam_tokens.h \
 		pam_modutil_private.h pam_static_modules.h \
-		include/pam_cc_compat.h
+		include/pam_cc_compat.h include/pam_inline.h
 
 libpam_la_LDFLAGS = -no-undefined -version-info 83:1:83
 libpam_la_LIBADD = @LIBAUDIT@ $(LIBPRELUDE_LIBS) @LIBDL@
diff --git a/libpam/include/pam_cc_compat.h b/libpam/include/pam_cc_compat.h
index a4b84c6..1aaed77 100644
--- a/libpam/include/pam_cc_compat.h
+++ b/libpam/include/pam_cc_compat.h
@@ -32,4 +32,17 @@
 # define DIAG_POP_IGNORE_CAST_QUAL	/* empty */
 #endif
 
+/*
+ * Evaluates to
+ * 1, if the given two types are known to be the same
+ * 0, otherwise.
+ */
+#if PAM_GNUC_PREREQ(3, 0)
+# define PAM_IS_SAME_TYPE(x_, y_)					\
+	__builtin_types_compatible_p(__typeof__(x_), __typeof__(y_))
+#else
+/* Cannot tell whether these types are the same.  */
+# define PAM_IS_SAME_TYPE(x_, y_)	0
+#endif
+
 #endif /* PAM_CC_COMPAT_H */
diff --git a/libpam/include/pam_inline.h b/libpam/include/pam_inline.h
new file mode 100644
index 0000000..d93f847
--- /dev/null
+++ b/libpam/include/pam_inline.h
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2020 Dmitry V. Levin <ldv@altlinux.org>
+ *
+ * Handy inline functions and macros providing some convenient functionality
+ * to libpam and its modules.
+ */
+
+#ifndef PAM_INLINE_H
+#define PAM_INLINE_H
+
+#include "pam_cc_compat.h"
+
+/*
+ * Evaluates to
+ * - a syntax error if the argument is 0,
+ * 0, otherwise.
+ */
+#define PAM_FAIL_BUILD_ON_ZERO(e_)	(sizeof(int[-1 + 2 * !!(e_)]) * 0)
+
+/*
+ * Evaluates to
+ * 1, if the given type is known to be a non-array type
+ * 0, otherwise.
+ */
+#define PAM_IS_NOT_ARRAY(a_)		PAM_IS_SAME_TYPE((a_), &(a_)[0])
+
+/*
+ * Evaluates to
+ * - a syntax error if the argument is not an array,
+ * 0, otherwise.
+ */
+#define PAM_MUST_BE_ARRAY(a_)		PAM_FAIL_BUILD_ON_ZERO(!PAM_IS_NOT_ARRAY(a_))
+
+/* Evaluates to the number of elements in the specified array.  */
+#define PAM_ARRAY_SIZE(a_)		(sizeof(a_) / sizeof((a_)[0]) + PAM_MUST_BE_ARRAY(a_))
+
+#endif /* PAM_INLINE_H */
diff --git a/modules/pam_namespace/pam_namespace.c b/modules/pam_namespace/pam_namespace.c
index ca6e22a..e1d8b9a 100644
--- a/modules/pam_namespace/pam_namespace.c
+++ b/modules/pam_namespace/pam_namespace.c
@@ -35,6 +35,7 @@
 #define _ATFILE_SOURCE
 
 #include "pam_cc_compat.h"
+#include "pam_inline.h"
 #include "pam_namespace.h"
 #include "argv_parse.h"
 
-- 
2.47.1

